src/engine/renderer/DetectGLVendors.h

```diff
@@ -0,0 +1,300 @@
+	  notice, this list of conditions and the following disclaimer.
+	* Redistributions in binary form must reproduce the above copyright
+	  notice, this list of conditions and the following disclaimer in the
+	  documentation and/or other materials provided with the distribution.
+	* Neither the name of the Daemon developers nor the
+	  names of its contributors may be used to endorse or promote products
+	  derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL DAEMON DEVELOPERS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+===========================================================================
+*/
+
+// DetectGLVendors.h
+
+#ifndef __DETECT_OPENGL_VENDORS
+#define __DETECT_OPENGL_VENDORS
+
+#include "common/Assert.h"
+#include "common/Common.h"
+
+enum class glHardwareVendor_t
+{
+	UNKNOWN,
+	// Assumed to be slow (running on CPU).
+	SOFTWARE,
+	/* Assumed to be fast (running on GPU through a translation, like an OpenGL over Vulkan
+	driver, or an OpenGL virtualization driver sending commands to an hypervisor. */
+	TRANSLATION,
+	// Assumed to be fast (running on GPU directly).
+	ATI,
+	BROADCOM,
+	INTEL,
+	LIMA,
+	NVIDIA,
+};
+
+inline std::string GetGLHardwareVendorName( glHardwareVendor_t hardwareVendor )
+{
+	static const std::string hardwareVendorNames[] = {
+		"Unknown",
+		"Software rasterizer",
+		"Translator",
+		"AMD/ATI",
+		"Broadcom",
+		"Intel",
+		"Lima",
+		"Nvidia",
+	};
+
+	if ( Util::ordinal(hardwareVendor) < 0
+		|| Util::ordinal(hardwareVendor) >= ( sizeof(hardwareVendorNames) / sizeof(hardwareVendorNames[0]) ) )
+	{
+		ASSERT_UNREACHABLE();
+		return "OutOfRange";
+	}
+
+	return hardwareVendorNames[ Util::ordinal(hardwareVendor) ];
+}
+
+enum class glDriverVendor_t
+{
+	UNKNOWN,
+	ATI,
+	INTEL,
+	MESA,
+	NVIDIA,
+};
+
+inline std::string GetGLDriverVendorName( glDriverVendor_t driverVendor )
+{
+	static const std::string driverVendorNames[] = {
+		"Unknown",
+		"AMD/ATI",
+		"Intel",
+		"Mesa",
+		"Nvidia",
+	};
+
+	if( Util::ordinal(driverVendor) < 0
+		|| Util::ordinal(driverVendor) >= ( sizeof(driverVendorNames) / sizeof(driverVendorNames[0]) ) )
+	{
+		ASSERT_UNREACHABLE();
+		return "OutOfRange";
+	}
+
+	return driverVendorNames[ Util::ordinal(driverVendor) ];
+}
+
+inline void DetectGLVendors(
```

Comment:
Why don't you put the functions in a .cpp file?

Reply:
I prefer to have the `*VendorNames` string arrays to be stored next to their related `gl*Vendor_t` enum classes for easier maintenance and to reduce risk of introducing discrepancies, meaning the only remaining non-inline function would be `DetectGLVendors` which is actually used only once by a single file… So splitting the file into `.h/.cpp` looks overkill to me.

Reply:
The more relevant question is not now many times it's used, but how many times it's included (the answer is in most client files this is included into client.h). Having a one-to-one relationship between .h and .cpp files is what more organized projects usually do

Reply:
I moved every functions to a `.cpp` file. I redesigned the things a bit to add many assertions to help keep in sync the string arrays with the enums.